### PR TITLE
Update the cmake build configuration for AppleClang compiler

### PR DESCRIPTION
**Pull request by JerryShih**
This pr try to merge the https://github.com/pytorch/pytorch/pull/11563 again and fix the linking error in https://github.com/pytorch/pytorch/pull/14837.

### CODE REVIEW

**Comment by ezyang** [comment]
Hmm... GitHub makes this annoying to do, but I'd kind of like it if we could land these patches individually :>

**Comment by JerryShih** [comment]
@ezyang
Sorry, I can't catch you. Do you mean I should squash all patches for reviewing?

**Comment by ezyang** [comment]
Vice versa actually. But we'll make do.

**Comment by JerryShih** [comment]
https://github.com/pytorch/pytorch/pull/15820/commits

I try to split the patches to make each commits clearly. Hope these patches could be reviewed easily.

**Comment by ezyang** [comment]
Hi @JerryShih, unfortunately, when I build your branch on CentOS, OpenMP is still disabled. More details coming soon.

**Comment by JerryShih** [comment]
> Hi @JerryShih, unfortunately, when I build your branch on CentOS, OpenMP is still disabled. More details coming soon.

@ezyang 
That's odd. I think I just setup the special option/setting for "AppleClang" compiler only. Could you make sure the master code-base has openmp before on CentOS?

This commit [[1]](https://github.com/pytorch/pytorch/pull/15820/commits/0567db3744cc92fa79825ce1a688ea10cccb4136) show the explicit message when the cmake finds openmp successfully for "pytorch". Maybe we could land this trivial patch first to get the openmp status in ci log.

**Comment by ezyang** [comment]
Here's the logs from the build: https://gist.github.com/ezyang/c49f798bd1955fdac2bd766b2caeed9a

I confirmed that OpenMP was "disabled" by running

```
>>> import torch
torch.get_num_threads()                                                             
1
```

I'll try and land your first commit, as you suggest.

**Comment by JerryShih** [comment]
> Here's the logs from the build: https://gist.github.com/ezyang/c49f798bd1955fdac2bd766b2caeed9a
> 
> I confirmed that OpenMP was "disabled" by running
> 
> ```
> >>> import torch
> torch.get_num_threads()                                                             
> 1
> ```
> I'll try and land your first commit, as you suggest.

@ezyang 
Hold on. I found the message "pytorch is compiling with OpenMP" in https://gist.github.com/ezyang/c49f798bd1955fdac2bd766b2caeed9a
I should trace the torch.get_num_threads(). I don't know the relationship between get_num_threads() and openmp.

**Comment by ezyang** [comment]
Yes, my other hypothesis is we might need to rebase to master.

**Comment by JerryShih** [comment]
@ezyang 
Ok, I know why you get 1 from torch.get_num_threads().

This python call will call to [[1]](https://github.com/pytorch/pytorch/blob/3b337e7892a7b54758f273477aa092053f02b410/aten/src/TH/THGeneral.cpp#L281),  but I just turn off the openmp for aten library in my patch set. You could test with "USE_OPENMP=ON" in caffe2 cmake [[2]](https://github.com/pytorch/pytorch/blob/3b337e7892a7b54758f273477aa092053f02b410/CMakeLists.txt#L128). If this build config could resolve the problem, I should update the aten cmake config logic.

**Comment by ezyang** [comment]
I tested just your commit df7648513cb947ca61bba2cf7308258abdd4a805 and it reported the correct number of threads. So there is definitely something later that is causing the problem.

**Comment by ezyang** [comment]
I'm merging the first commit in your series at #16506

**Comment by JerryShih** [comment]
@ezyang 
Here is the rebase one.
I also turn on the openmp by default in [[1]](https://github.com/pytorch/pytorch/pull/15820/commits/6cc185644714888ca1a4e92410e349bacd001f26).

Please try to test with:
> import torch
torch.get_num_threads()                                                             
1

I don't have a good solution to test the openmp status in pytorch.
In my mind, I could use "add_test()" to add a python openmp test.
After the pytorch build, call "make test"(or ninja test) to trigger this python script.

File: torch/CMakeLists.txt [hunk 1]
```diff
@@ -323,11 +325,18 @@ target_link_libraries(torch caffe2_library)
...
+  target_compile_options(torch INTERFACE ${OpenMP_CXX_FLAGS})
+  if (CMAKE_CXX_COMPILER_ID MATCHES "AppleClang")
+    #AppleClang 7+ should link the separated OpenMP lib.
+    #https://github.com/Kitware/CMake/blob/42212f7539040139ecec092547b7d58ef12a4d72/Modules/FindOpenMP.cmake#L252
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "7.0" OR
+        CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "7.0")
+      target_link_libraries(torch ${OpenMP_libomp_LIBRARY})
+    endif()
+  else()
+    target_link_libraries(torch -fopenmp)
```
Comment (SsnL) [inline torch/CMakeLists.txt]: I think in general blindly linking `-fopenmp` is a bad idea. Considering the problem with `OpenMP_libomp_LIBRARY` being not available in 3.5 cmake, I would suggest port newer `FindOpenMP.cmake` over, and link against `OpenMP_<lang>_LIBRARIES`.

File: tools/build_pytorch_libs.py [hunk 2]
```diff
@@ -181,7 +181,15 @@ def run_cmake(version,
         USE_REDIS=os.getenv('USE_REDIS'),
         USE_GLOG=os.getenv('USE_GLOG'),
         USE_GFLAGS=os.getenv('USE_GFLAGS'),
-        WERROR=os.getenv('WERROR'))
+        WERROR=os.getenv('WERROR'),
+        # openmp cmake flag
+        # If we setup MACOSX_DEPLOYMENT_TARGET in mac, we should also set the
+        # libomp library flags explicitly.
+        OpenMP_C_FLAGS=os.getenv('OpenMP_C_FLAGS'),
```
Comment (SsnL) [inline tools/build_pytorch_libs.py]: Are these really needed? Can't you instead use general purpose flags like `CFLAGS` and `LDFLAGS`?

File: tools/build_pytorch_libs.py [hunk 3]
```diff
@@ -181,7 +181,15 @@ def run_cmake(version,
...
+        OpenMP_CXX_FLAGS=os.getenv('OpenMP_CXX_FLAGS'),
+        OpenMP_C_LIB_NAMES=os.getenv('OpenMP_C_LIB_NAMES'),
+        OpenMP_CXX_LIB_NAMES=os.getenv('OpenMP_CXX_LIB_NAMES'),
+        OpenMP_omp_LIBRARY=os.getenv('OpenMP_omp_LIBRARY'))
```
Comment (SsnL) [inline tools/build_pytorch_libs.py]: This is not used anywhere. `OpenMP_libomp_LIBRARY` you used below is actually set by `FindOpenMP.cmake`.

**Review by ssnl — COMMENTED** [review]
Thanks for doing this in a generally better strategy than what I did :). I have a couple of comments, with the main suggestion being to eliminate the unnecessary extra env flags introduced (i.e., `OpenMP_xxxx`). No user would be able to find them out unless they want to read our CI scripts. Moreover, I think OpenMP is more likely & easier to be used by user if the mechanism can be replaced by 

1. smarter cmake code that test `AppleClang > 7.0` and then set `-Xpreprocessor -fopenmp`, and
2. general purpose flags like `CFLAGS` and `LDFLAG`.

Alternatively, we can also port the new `FindOpenMP.cmake` over, which supports `AppleClang` and discovers `brew` installed locations as long as `/usr/local/lib` is in the search path. :)

**Comment by JerryShih** [comment]
@SsnL 
Thank you, I'm trying to have the custom FindOpenMP module in "pytorch/cmake/Modules".

File: torch/CMakeLists.txt (hunk 1 above)
Comment (JerryShih) [reply torch/CMakeLists.txt]: Update to use ${OpenMP_CXX_LIBRARIES} now.

File: caffe2/CMakeLists.txt [hunk 4]
```diff
@@ -224,6 +224,19 @@ if(NOT BUILD_ATEN_ONLY)
   endif()
 endif()
 
+if (USE_OPENMP AND OPENMP_FOUND)
+  if (CMAKE_CXX_COMPILER_ID MATCHES "AppleClang")
+    #AppleClang 7+ should link the separated OpenMP lib.
+    #https://github.com/Kitware/CMake/blob/42212f7539040139ecec092547b7d58ef12a4d72/Modules/FindOpenMP.cmake#L252
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "7.0" OR
+        CMAKE_CXX_COMPILER_VERSION VERSION_GREATER "7.0")
+      target_link_libraries(caffe2 PRIVATE ${OpenMP_libomp_LIBRARY})
```
[the remark this answers is not in the document]
Comment (JerryShih) [reply caffe2/CMakeLists.txt]: Update to use ${OpenMP_CXX_LIBRARIES} now.

File: tools/build_pytorch_libs.py (hunk 2 above)
Comment (JerryShih) [reply tools/build_pytorch_libs.py]: The change in build_pytorch_libs.py are removed now.

File: tools/build_pytorch_libs.py (hunk 3 above)
Comment (JerryShih) [reply tools/build_pytorch_libs.py]: These changes are reverted.

**Comment by JerryShih** [comment]
@ezyang @SsnL 

Finally, here is the new one.

What I have done here are:
1. add a custom findopenmp module copied from cmake v3.13.
It will check "brew libomp" default install path if we setup MACOSX_DEPLOYMENT_TARGET flag.

2. **use ${OpenMP_CXX_LIBRARIES} in linking setting.**
**This is the risky one. It should be tested with ppc64le platform. I remember that I broke the ppc64le build in my previous patch(https://github.com/pytorch/pytorch/pull/11563). I don't have that platform locally, so I can't test for that platform.**

3. move the openmp checking code from aten code block to "USE_OPENMP" checking block

4. add the hardcoded linking setting for pthread in asan build

5. install libomp from brew for all macos related test builds.

**Comment by SsnL** [comment]
@JerryShih Thanks. The CI failures are due to the tricky case where MKL links a conda `libiomp5` and `FindOpenMP` links a brew one. I'm trying to fix this, and push to your branch.

**Comment by SsnL** [comment]
Locally I've verified that this allows me to use multithreaded reductions like `sum` on CPU.

**Comment by SsnL** [comment]
Will the bad news is that it fails about every CI build with this sccache error:
```
Feb 02 03:26:16 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/ccAtd869.s: Assembler messages:\n/tmp/ccAtd869.s:72: Error: operand type mismatch for `vxorps\'\n/tmp/ccAtd869.s:76: Error: operand type mismatch for `vandps\'\n/tmp/ccAtd869.s:110: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:113: Error: no such instruction: `kmovb %eax,%k3\'\n/tmp/ccAtd869.s:115: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:116: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:117: Error: operand size mismatch for `vmovdqa64\'\n/tmp/ccAtd869.s:118: Error: operand type mismatch for `vpxord\'\n/tmp/ccAtd869.s:120: Error: no such instruction: `kmovb %eax,%k4\'\n/tmp/ccAtd869.s:121: Error: operand type mismatch for `vpabsq\'\n/tmp/ccAtd869.s:122: Error: operand size mismatch for `vmovdqa64\'\n" }
```
Any ideas? @ezyang @yf225

**Comment by SsnL** [comment]
Oh I can actually reproduce this on ubuntu. Not on mac tho.

File: cmake/Modules/FindMKL.cmake [hunk 5]
```diff
@@ -116,30 +120,52 @@ MACRO(CHECK_ALL_LIBRARIES LIBRARIES _name _list _flags)
...
     IF(_libraries_work)
-      IF(${_library} STREQUAL "gomp")
-          FIND_PACKAGE(OpenMP)
+      IF(${_library} MATCHES "omp")
+        IF(_openmp_type)
+          MESSAGE(FATAL_ERROR "More than one OpenMP libraries appear in the MKL test: ${_list}")
+        ELSEIF(${_library} MATCHES "gomp")
+          SET(_openmp_type "GNU")
+          # Use FindOpenMP to find gomp
+          FIND_PACKAGE(OpenMP QUIET)
```
Comment (SsnL) [inline cmake/Modules/FindMKL.cmake]: oh... findmkl and findomp probably should not recursively call into each other :P

**Comment by ezyang** [comment]
Oh! When I cleaned and rebuilt the problem went away lol.

**Comment by SsnL** [comment]
Pro tip: use ‘—cmake’ when there are canker changes :D

**Pull request closed without merging**